rotkehlchen/db/dbhandler.py

```diff
@@ -2813,19 +2841,40 @@ def query_timed_balances(
         cursor = self.conn.cursor()
         results = cursor.execute(querystr, bindings)
         results = results.fetchall()
-        balances = []
+        balances: List[SingleDBAssetBalance] = []
+        # In the case of eth and ETH2 we have to agregate their value
+        last_eth_timestamp = 0
+        last_eth_and_eth2_assets: List[SingleDBAssetBalance] = []
+        last_eth_and_eth2_liabilities: List[SingleDBAssetBalance] = []
+
         results_length = len(results)
         for idx, result in enumerate(results):
             entry_time = result[0]
             category = BalanceType.deserialize_from_db(result[3])
-            balances.append(
-                SingleDBAssetBalance(
-                    time=entry_time,
-                    amount=result[1],
-                    usd_value=result[2],
-                    category=category,
-                ),
+            if settings.treat_eth2_as_eth is True and entry_time > last_eth_timestamp:
```

Comment:
This entire logic here is terribly complicated and convoluted. I spent about 10 minutes trying to understand what you are trying to do and still not sure exactly why you went with such a complicated logic.

Why do you have to call `aggregate_eth_asset_balances` in every iteration? Why do you do modify the balances list in place?

What are you trying to achieve here and why is the code so complicated?

Why can't you just let the code as it was before and if setting is on have a post-processing function that will sum up all `SingleAssetBalance` entries with the same Asset and same timestamp? This will solve the problem in a much more elegant way and also be applicable for any assets not just eth and eth2.

Reply:
The logic there does that. Combine the balances with same asset, same timestamp and same type (liability, asset). Do you mean to do it at a higher level?

Reply:
The iteration is over each timestamp and what it does if check if there are pending eth2 and eth balances to merge

Reply:
The logic is doing this in the most confusing way possible. Please rewrite with the way I explained in my last paragraph.